File: incre-tests/fusion/shortcut/page1.f
Inductive List = nil Unit | cons {Int, List};
Inductive BoolList = bnil Unit | bcons {Bool, BoolList};

map = \f: Int -> Bool. fix (
  \g: List -> Compress BoolList. \xs: List.
  match xs with
    nil _ -> bnil unit
  | cons {h, t} -> bcons {f h, g t}
  end
);

p = \x: Int. >= 0 x;

all = fix (
  \f: BoolList -> Bool. \xs: BoolList.
  match xs with
    bnil _ -> true
  | bcons {h, t} -> and h (f t)
  end
);

main = \xs: List. all (map p xs);

File: ml-benchmark/fusion/shortcut/page1.ml
type list =
  | Nil
  | Cons of int * list

type bool_list =
  | BNil
  | BCons of bool * bool_list

let rec map f xs =
  match xs with
  | Nil -> BNil
  | Cons (h, t) ->
      let bh = f h in
      let bt = map f t in
      BCons (bh, bt)

let p x = x >= 0

let rec all xs =
  match xs with
  | BNil -> true
  | BCons (h, t) -> h && all t

let program xs = all (map p xs)